c/src/ml-api-common.c

```diff
@@ -1287,25 +1287,117 @@ ml_strerror (int errnum)
  * @brief Internal function for destroy value of option table
  */
 static void
-_ml_option_value_free (gpointer data)
+_ml_info_value_free (gpointer data)
 {
-  ml_option_value_s *_value;
+  ml_info_value_s *_value;
 
-  _value = (ml_option_value_s *) data;
+  _value = (ml_info_value_s *) data;
   if (_value) {
     if (_value->destroy)
       _value->destroy (_value->value);
     g_free (_value);
   }
 }
 
+/**
+ * @brief Internal function for create ml_info
+ */
+static ml_info_s *
+_ml_info_create (ml_info_type_e type)
+{
+  ml_info_s *info;
+
+  info = g_new0 (ml_info_s, 1);
+  if (info == NULL) {
+    _ml_error_report
+        ("Failed to allocate memory for the ml_info. Out of memory?");
+    return NULL;
+  }
+
+  info->type = type;
+  info->table =
+      g_hash_table_new_full (g_str_hash, g_str_equal, g_free,
+      _ml_info_value_free);
+  if (info->table == NULL) {
+    _ml_error_report
+        ("Failed to allocate memory for the table of ml_info. Out of memory?");
+    g_free (info);
+    return NULL;
+  }
+
+  return info;
+}
+
+/**
+ * @brief Internal function for destroy ml_info
+ */
+static void
+_ml_info_destroy (ml_info_s * info)
+{
+  if (!info)
+    return;
+
+  if (info->table)
+    g_hash_table_destroy (info->table);
+
+  g_free (info);
+  return;
+}
+
+/**
+ * @brief Internal function for set value of given ml_info
+ */
+static int
+_ml_info_set_value (ml_info_s * info, const char *key, void *value,
+    ml_data_destroy_cb destroy)
+{
+  ml_info_value_s *info_value;
+
+  if (!info || !key || !value)
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'info', 'key' or 'value' is NULL. It should be a valid ml_info, key and value.");
+
+  info_value = g_new0 (ml_info_value_s, 1);
+  if (!info_value)
+    _ml_error_report_return (ML_ERROR_OUT_OF_MEMORY,
+        "Failed to allocate memory for the info value. Out of memory?");
+
+  info_value->value = value;
+  info_value->destroy = destroy;
+  g_hash_table_insert (info->table, g_strdup (key), (gpointer) info_value);
+
+  return ML_ERROR_NONE;
+}
+
+/**
+ * @brief Internal function for get value of given ml_info
+ */
+static int
+_ml_info_get_value (ml_info_s * info, const char *key, void **value)
+{
+  ml_info_value_s *info_value;
+
+  if (!info || !key || !value)
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'info', 'key' or 'value' is NULL. It should be a valid ml_info, key and value.");
+
+  info_value = g_hash_table_lookup (info->table, key);
+  if (!info_value)
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
```

Comment:
Let's reduce the error message, sometimes business division complains unnecessary error logs.
```if (!info_value) return ML_ERROR_INVALID_PARAMETER;```

Reply:
Thanks, I updated the code PTAL.